src/material/radio/radio.ts

```diff
@@ -695,6 +700,31 @@ export class MatRadioButton implements OnInit, AfterViewInit, DoCheck, OnDestroy
       if (input) {
         input.setAttribute('tabindex', value + '');
         this._previousTabIndex = value;
+        // Wait for any pending tabindex changes to be applied
+        afterNextRender(
+          () => {
+            queueMicrotask(() => {
+              // The radio group uses a "selection follows focus" pattern for tab management, so if this
+              // radio button is currently focused and another radio button in the group becomes
+              // selected, we should move focus to the newly selected radio button to maintain
+              // consistency between the focused and selected states.
+              if (
+                group &&
+                group.selected &&
+                group.selected !== this &&
+                document.activeElement === input
+              ) {
+                group.selected?._inputElement.nativeElement.focus();
+                // If this radio button still has focus, the selected one must be disabled. In this
+                // case the radio group as a whole should lose focus.
+                if (document.activeElement === input) {
+                  this._inputElement.nativeElement.blur();
```

Comment:
What will receive focus after this is called? Seems like we should be pushing focus to something nearby

Reply:
Focus will shift to the body (this is the same thing that would happen if the focused element became disabled, as shown in this example: https://jsfiddle.net/ogrsy2w0/). This case represents a very similar situation. Because the selected option is disabled, the other options are effectively disabled. You can't click any of them because that would change the selectedness of the disabled option.

Maybe we should actually take it a step further and instead of making them effectively disabled, just make all of the radios in the group actually disabled. Then I wouldn't even need the blur logic, I could just let the browser do its thing. WDYT?

Reply:
Yeah this seems like a weird experience to have a selected radio button disabled, while the others aren't. It seems like the whole group should be disabled in that case. 

However, would that introduce any weird edge cases where we set the group as disabled? What if the radio has its selected changed programmatically to a non-disabled button? We'd want to make sure to undo that disabled group effect, which I assume means we'd need some additional state information to understand the situation.

I assume this is a really unique case anyways, where a focused radio button becomes disabled.

Another alternative is that we quickly disable the group and then immediately un-disable it so that the browser moves the focus away.

Reply:
Yeah that's a good point, I feel like extending this to handle the concept of "disabled due to selected item being disabled" is not trivial, and probably easy to mess up. How do you feel about submitting this PR to at least handle the focus state correctly, and just leaving a TODO to deal with communicating the disabled state visually / to the screen reader?

This is a great example of a common pattern where its really tricky to get all the interaction edge cases correct, and it would be amazing to have some shared code for it. This same issue will apply to any component that implements this kind of mutually exclusive selection mechanism (e.g. radio, single-selection listbox, single-selection button toggle, single-selection chips, etc.)